### DISCUSSION TITLE
How do you approach passing around ID's for resources without it becoming a hot mess?

### DISCUSSION THREAD

**Discussion by sheldonhull** [pulumi/pulumi#9205] [Pulumi in Practice] [answered]
## Avoid setting metadata name (physical) to preserve unique naming

I've learned that not setting the metadata name in Kubernetes resources can help ensure each resource has a unique physical name. However, things get a bit confusing when I think through how to do this properly in a multi-cluster and beyond context. 

For example, if I want to have a unique URN I need the logical name to be unique, and to not override the metadata name when possible. 

## Challenges With StringOutput & String

However, the logical name for most resources is a string type, and the outputs for Pulumi are StringOutput. When I examined how to resolve, I realize that we get into the challenges of the dependency tree and asynchronous code execution in the Pulumi engine. 

I can't guarantee in this plan that I will apply "random string" first, use that output and then combine this StringOutput type with a normal string type to create a unique logical name.

My attempts to resolve this with Go remind me of C# delegate functions and the resulting `ApplyT` style expressions are not very understandable or have many Go based examples to figure this out.

## Using A Random Resource and Order of Apply Operations

Using a resource like random. I'll create a separate discussion to get some insight on this.

- Created: https://github.com/pulumi/pulumi/discussions/9207

## How to Pass the Outputs Without a Huge Mess

To simplify maintenance I've broken up my various resources into their own internal package. I then need to consume each resources id to pass to the other functions. However, how do I do this without it becoming a huge mess?

```go
pulumi/myapp
├── Pulumi.dev.yaml
├── Pulumi.yaml
├── README.md
├── go.mod
├── go.sum
├── internal
│   ├── configmap
│   │   └── configmap.go
│   ├── deployment
│   │   └── deployment.go  👈 Especially problematic, needs configmap, role/binding/service,serviceaccount outputs before it can run
│   ├── disruption
│   │   └── disruption.go
│   ├── hpa
│   │   └── hpa.go
│   ├── ingress
│   │   └── ingress.go
│   ├── role
│   │   └── role.go
│   ├── rolebinding
│   │   └── rolebinding.go
│   ├── service
│   │   └── service.go
│   └── serviceaccount
│       └── serviceaccount.go
├── main.go 
├── pkg
│   └── config
│       └── config.go   👈 Load all my config into structs and use methods for naming convention/string formatting
```

Is the only viable option passing a bunch of individual parameters, or is there something perhaps I can leverage with context, exports, or something else to get the name resolved and pass it in a struct without conflicts?

Example:

```go
func ConfigMap(
	ctx *pulumi.Context,
	prov *kubernetes.Provider,
	configData *planconfig.SvcConfig,
	appConfig *planconfig.AppConfig,
	sharedConfig *planconfig.SharedConfig,
) (cfgMap *corev1.ConfigMap, err error) {} // 👈 see output type of function
vs
func ConfigMap(
	ctx *pulumi.Context,
	prov *kubernetes.Provider,
	configData *planconfig.SvcConfig,
	appConfig *planconfig.AppConfig,
	sharedConfig *planconfig.SharedConfig,
) (cfgMapName string, err error) { // 👈 see output type of function
```

If I cfgMap, err = corev1.NewConfigMap( and then return cfgMap then other functions need that type and to resolve Elem().Name or whatever it is. I'd like to simplify.

Maybe

```go
// This struct or equivalent has to be able to be safe to use with those asynchronous jobs running. 
type CombinedCreatedResources struct { 
DeploymentName string
ConfigMapName string
...etc
}
```

## Current WorkAround

My current work around was actually just using `petname` cli and adding a random petname to the logical name for resources. This is manually put into the configuration yaml though, and definitely not what I want to be doing. I want to get rid of any overriding the actual name and use the generated name to allow for the full benefit of Pulumi.

**Comment by ghostsquad** [comment, marked as answer]
@sheldonhull it seems you are describing several problems. One, you want to deal with concrete strings _and_ asynchronous strings in a unified way, which is what ApplyT is for. 

For my own sanity, I've found that if I write my own resources, which _wrap_ the pulumi resources, then I have much better control over the use of `string` and other concrete values as opposed to the "async" values in pulumi.

```
import (
    pCoreV1 "github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes/core/v1"
)

type ConfigMap {
    pConfigMap *pcorev1.ConfigMap
    ...
}

func NewConfigMap(...) (*ConfigMap, error) {

}
```

For your second problem `How to Pass the Outputs Without a Huge Mess`. For kubernetes resources, MOST values can be obtained at compile time. Rarely will you need a value you can't predict or programmatically identify. An example of this may be a AWS Role ARN. You may think that it's easiest to use the pulumi.Role.ARN field (async), but if you are creating the role from strings, you can store those strings in a different struct (your wrapper), and pass that around.

**Reply by sheldonhull** [reply]
@ghostsquad Take your example above, how would you use ApplyT to obtain the generated ID  so I could pass it to a deployment?

The approach you have above is very similar to what I did except I just had each resource as a separate package with their own function. You are wrapping up there pointer in your own struct and using methods which makes a perfect sense considering we care about the state.

Any example on how you take it to the next level by applying and returning a string value that could then be used by another resource would be really helpful.

**Reply by ghostsquad** [reply]
What value are you looking for? You said the "generated id". Can you clarify this please?
